Unit test ssa_exprt

romainbrenguier opened the pull request:
This adds unit test for the ssa_exprt class ~and clean-up some things I noticed while writing the tests, in particular the constructor cannot take any kind of `exprt` but a very specific class of these, which I documented.~


<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Add unit test for ssa_exprt

Makes it possible to test the class in isolation.

romainbrenguier committed:
Unit test for ssa_exprt::set_expression

romainbrenguier committed:
Unit test for ssa_exprt::get_object_name

This gives examples of what get_object_name is supposed to do.

romainbrenguier committed:
Unit test for ssa_exprt::get_l1_object

This tests the behavior of the method.

romainbrenguier committed:
Check for get_identifier in unit test

This checks that operations mutating the ssa_exprt give it a correct identifier.

romainbrenguier committed:
Check for l1 object identifier in unit test

This checks that methods mutating ssa_exprt, correctly set the l1 object
identifier.

tautschnig commented on unit/util/ssa_expr.cpp:
```diff
@@ -48,3 +48,33 @@ TEST_CASE("Set level", "[unit][util][ssa_expr]")
+      }
+    }
+  }
+}
```
Github complains about a missing newline (which is fixed in "Unit test for the ssa_exprt constructor")

tautschnig reviewed (changes requested):
This strictly requires a performance evaluation as various methods can no longer be inlined (unless we start using LTO).

romainbrenguier commented:
I have extracted commits there https://github.com/diffblue/cbmc/pull/4736, to only leave unit tests here. So that performance can be evaluated independently and doesn't block merging the tests.

romainbrenguier commented:
@tautschnig this is ready for a second look

tautschnig commented:
@romainbrenguier clang-format is unhappy.

romainbrenguier committed:
Unit test for ssa_exprt constructor

This check what happens when the constructor is called on an expression
containing symbol, member access and array access.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fb7abbd86e7dad8a585dd6ec9f7e5f853fe71c59).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113797074
